tests/SpecTests/PrimaryStepDownSpecTest.php

```diff
@@ -0,0 +1,251 @@
+            $this->insertDocuments(1);
+        } catch (BulkWriteException $e) {
+            // Verify that the insert failed with an operation failure with 10107 code.
+            $this->assertSame(91, $e->getCode());
+        }
+
+        // Verify that the connection pool has been cleared
+        $this->assertSame($totalConnectionsCreated + 1, $this->getTotalConnectionsCreated());
+    }
+
+    public function testInterruptedAtShutdownResetConnectionPool()
+    {
+        $runOn = [(object) ['maxServerVersion' => '4.0.999', 'topology' => [self::TOPOLOGY_REPLICASET]]];
+        $this->checkServerRequirements($runOn);
+
+        // Set a fail point
+        $this->configureFailPoint([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'failCommands' => ['insert'],
+                'errorCode' => 11600,
+            ],
+        ]);
+
+        $totalConnectionsCreated = $this->getTotalConnectionsCreated();
+
+        // Execute an insert into the test collection of a {test: 1} document.
+        try {
+            $this->insertDocuments(1);
+        } catch (BulkWriteException $e) {
+            // Verify that the insert failed with an operation failure with 10107 code.
+            $this->assertSame(11600, $e->getCode());
+        }
+
+        // Verify that the connection pool has been cleared
+        $this->assertSame($totalConnectionsCreated + 1, $this->getTotalConnectionsCreated());
+    }
+
+    private function insertDocuments($count)
+    {
+        $operations = [];
+
+        for ($i = 1; $i <= $count; $i++) {
+            $operations[] = [
+                BulkWrite::INSERT_ONE => [['test' => $i]],
+            ];
+        }
+
+        return $this->collection->bulkWrite($operations, ['writeConcern' => new WriteConcern('majority')]);
+    }
+
+    private function dropAndRecreateCollection()
+    {
+        $this->client->selectCollection($this->getDatabaseName(), $this->getCollectionName())->drop();
+        $this->client->selectDatabase($this->getDatabaseName())->command(['create' => $this->getCollectionName()]);
+    }
+
+    private function getTotalConnectionsCreated()
+    {
+        $cursor = $this->client->getManager()->executeCommand(
+            $this->getDatabaseName(),
+            new Command(['serverStatus' => 1]),
+            new ReadPreference(ReadPreference::RP_PRIMARY)
+        );
+
+        $cursor->setTypeMap(['root' => 'array', 'document' => 'array']);
+        $document = current($cursor->toArray());
+
+        if (isset($document['connections'], $document['connections']['totalCreated'])) {
+            return (int) $document['connections']['totalCreated'];
+        }
+
+        throw new UnexpectedValueException('Could not determine number of total connections');
+    }
+
+    private function waitForMasterReelection()
+    {
+        for ($i = 0; $i < 10; $i++) {
+            try {
+                $this->insertDocuments(1);
```

Comment:
Alternatively, could we just invoke server selection until a primary is returned?

https://github.com/mongodb/specifications/tree/master/source/connections-survive-step-down/tests#do-we-need-to-wait-for-re-election-after-the-first-test suggests that might be sufficient.

Also, I would suggest using a sleep of 0.5 seconds, as that is the non-configurable [minHeartbeatFrequencyMS](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#minheartbeatfrequencyms) value and theoretically smallest interval that a single-threaded driver would pick up topology changes.

Reply:
I've previously called `Manager::selectServer()` to fetch a server, but this always returns the previous primary even when the replica set does not have a primary yet. I'm not sure why libmongoc isn't picking up on the topology change. Even calling `Manager::selectServer()` directly after the `insert` failed with a `not master` error returns the same server, with it still identifying as a primary. This also happens when sleeping for a while to ensure `minHeartBeatFrequencyMS` passed. We should definitely investigate this to ensure we don't miss any topology changes.

Reply:
@alcaeus: The behavior you're describing would seem to conflict with ["not master" and "node is recovering"](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#not-master-and-node-is-recovering) in the SDAM spec.

> When the client sees a "not master" or "node is recovering" error it MUST replace the server's description with a default ServerDescription of type Unknown.
> ...
> For single-threaded clients, in the case of a "not master" or "node is shutting down" error, the client MUST mark the topology as "stale" so the next server selection scans all servers.

Feel free to open a CDRIVER ticket so we can look into that further.

Reply:
Correction: this was a bit of a failure on my part. The initial bulkWrite after a stepdown fails with a `not master` error, which correctly invalidates the topology and causes subsequent calls to `selectServer` to fail with a `ConnectionTimeoutException` until a new master has been elected. This slipped through the cracks since this test actually opens two connections to the server: one by instantiating the client itself, the other by `MongoDB\Tests\FunctionalTestCase` creating a new `Manager` instance in its own `setUp` method. Calling `MongoDB\Tests\FunctionalTestCase::getPrimaryServer` always uses this connection, which is not aware of the topology change since no writes are sent there.

I can change the logic in question, but this still needs an initial write operation on the old primary to get libmongoc to invalidate the topology and check for a new master. After this initial failure we can then perform server selection until a primary is returned.

One more thing to note is that I've observed elections taking quite long - sometimes up to 20 seconds. Is this expected behaviour?

Reply:
I now remember why the write is required. MongoDB 4.2 no longer drops the connection when executing `replSetStepDown` and instead returns a response to the client. This doesn't give SDAM a chance to change the server's state to unknown. Since the connection remains active, you effectively need to execute a write operation to trigger a "not master" error (or otherwise coerce some error with a fail point or network timeout).

I actually wrote about this in [DOCSP-4920](https://jira.mongodb.org/browse/DOCSP-4920), as it came up while we were writing this spec.

If you'd like to make a contribution to the spec, I think we'd do well to mention this in the answer for [Do we need to wait for re-election after the first test?](https://github.com/mongodb/specifications/tree/master/source/connections-survive-step-down/tests#do-we-need-to-wait-for-re-election-after-the-first-test). It currently instructs implementations to invoke server selection until a primary is available; however, that obviously won't work for MongoDB 4.2+.

> One more thing to note is that I've observed elections taking quite long - sometimes up to 20 seconds. Is this expected behaviour?

That certainly seems high. You can try tuning [`settings.electionTimeoutMillis`](https://docs.mongodb.com/manual/reference/replica-configuration/#rsconf.settings.electionTimeoutMillis). [Factors and Conditions that Affect Elections](https://docs.mongodb.com/manual/core/replica-set-elections/#factors-and-conditions-that-affect-elections) may have additional pointers.

You can probably also shave some more time by reducing the current 5-second timeout for the [`replSetStepDown`](https://docs.mongodb.com/manual/reference/command/replSetStepDown/) command.

Reply:
> You can probably also shave some more time by reducing the current 5-second timeout for the [`replSetStepDown`](https://docs.mongodb.com/manual/reference/command/replSetStepDown/) command.

The 5 seconds are required by the spec, see https://github.com/mongodb/specifications/blob/master/source/connections-survive-step-down/tests/README.rst#getmore-iteration.